text/0156-columnar-encoding.md

```diff
@@ -0,0 +1,411 @@
+            "attribute",
+            DataType::Struct(vec![
+                Field::new("name", DataType::Utf8, false),
+                Field::new("value", DataType::Utf8, false),
+            ]),
+            true,
+        ))
+    ), true),
+])
+```
+
+#### OpenTelemetry traces to Arrow mapping
+
+This Arrow schema describes the representation of traces.
+
+Labels are mapped to a set of dedicated column scoped by the logical struct 'labels'. The list of labels can be easily
+determined from the schema by any participants. Metrics follow the same organization and are scoped by the logical
+struct 'metrics'.
+
+Attributes with a list of structs instead of a struct of fields (columns). This mapping is based on the assumption that
+the number of attributes can vary greatly from one log entry to another for the same log stream. Arrow encodes this type
+representation with an offsets buffer and a child/data array. If this assumption is not true, another option will be to
+declare one column per attribute with a validity bitmap enabled. **This should be validated by experimentation on
+realistic datasets.**
+
+Events and links follow the same representation used for the attributes.
+
+For more details on the Arrow Memory Layout see this [document](https://arrow.apache.org/docs/format/Columnar.html#variable-size-binary-layout).
+
+```rust
+// Trace schema declaration
+Schema::new(vec![
+    // Time range
+    Field::new("start_time_unix_nano", DataType::UInt64, false),
+    Field::new("end_time_unix_nano", DataType::UInt64, false),
+
+    Field::new("trace_id", DataType::Binary, false),
+    Field::new("span_id", DataType::Binary, false),
+    Field::new("trace_state", DataType::Utf8, true),
+    Field::new("parent_span_id", DataType::Binary, true),
+    Field::new("name", DataType::Utf8, false),
+    Field::new("kind", DataType::UInt8, false),
+
+    // Attributes
+    Field::new("attributes", DataType::List(
+        Box::new(Field::new(
+            "attribute",
+            DataType::Struct(vec![
+                Field::new("name", DataType::Utf8, false),
+                Field::new("value", DataType::Utf8, false),
+            ]),
+            true,
+        ))
+    ), true),
+
+    // Events
+    Field::new("events", DataType::List(
+        Box::new(Field::new(
+            "event",
+            DataType::Struct(vec![
+                Field::new("time_unix_nano", DataType::UInt64, false),
+                Field::new("name", DataType::Utf8, false),
+                Field::new("attributes", DataType::List(
+                    Box::new(Field::new(
+                        "attribute",
+                        DataType::Struct(vec![
+                            Field::new("name", DataType::Utf8, false),
+                            Field::new("value", DataType::Utf8, false),
+                        ]),
+                        true,
+                    ))
+                ), true),
+            ]),
+            true,
+        ))
+    ), true),
```

Comment:
Thought experiment:

I think you could express Span events in the OpenTelemetry logs to Arrow mapping shown above, and then a data processor could attach logs to spans in the form of events. 

In a multi-variate protocol, how do we justify the decision to emit one span with a list of associated events, rather than as a flat buffer of events?

Reply:
This representation has been updated.